threatgrid/ctia
# Adds asset-ref relation to Asset graphql schema

Just like every entity can relate to multiple [Feedback records](https://github.com/threatgrid/ctia/blob/master/src/ctia/entity/feedback/graphql_schemas.clj#L33), every Asset record can be associated with zero, one, or multiple AssetMappings and/or AssetProperties.

> **Epic** #
> Close https://github.com/threatgrid/iroh/issues/4521
> Related https://github.com/threatgrid/ctia/pull/1064

*This the second part of #1064* 

<a name="qa">[§](#qa)</a> QA
============================

1. Create a few `Assets` 
2. Create a few `AssetMappings` with `asset_ref` pointing to `Assets`, preferably multiple `AssetMappings` per single `Asset`. Note that you'd have to use `Asset` and `AssetMapping` POST requests separately. At the moment Bundle Importing Assets and related AssetMappings using transient IDs in asset_ref is [not implemented yet](https://github.com/threatgrid/iroh/issues/4751). That means you'd have to create an Asset (`POST /ctia/asset`), write down its ID; Create AssetMapping(s) (`POST /ctia/asset-mapping`) with `asset_ref` to be of that ID, and so on.
3. Similarly, create multiple `AssetProperties` per each `Asset` object
4. Test by sending GraphQL queries  to `/ctia/graphql` (for each created Asset) like this:
```
{
    "query": "query Asset { asset(id:\"http://localhost:3000/ctia/asset/asset-61c378d5-c38d-45e1-a6f0-590d44e9cac8\") {id asset_mappings{ totalCount nodes { asset_ref }}}}"
}
    
```
    
  That query should return a map such this:
   
    {
      "data": {
        "asset": {
          "id": "http://localhost:3000/ctia/asset/asset-61c378d5-c38d-45e1-a6f0-590d44e9cac8",
          "asset_mappings": {
            "totalCount": 1,
            "nodes": [
              {
                "asset_ref": "http://localhost:3000/ctia/asset/asset-61c378d5-c38d-45e1-a6f0-590d44e9cac8"
              }
            ]
          }
        }
      }
    }

5. Make sure that `totalCount` is matching the number of `AssetMappings` created per `Asset`
6. Make sure that all `asset_ref` fields have the same `AssetId`
7. You can test `AssetProperties` similarly:

```
{
    "query": "query Asset { asset(id:\"http://localhost:3000/ctia/asset/asset-61c378d5-c38d-45e1-a6f0-590d44e9cac8\") {id asset_properties{ totalCount nodes { asset_ref }}}}"
}
```

<a name="squashed-commits">[§](#squashed-commits)</a> Squashed Commits
======================================================================

5b7c4ee6 change rand-btw to take half-open interval
21934579 test: asset graphql asset_ref test
dc3aad56 fix resolvers for asset-mapping and asset-properties
253a2ed5 conjoin graphql queries for Asset, AssetMapping and AssetProperties
6b1901f6 fix for: `Verdict not found in schema`
3b645da8 adds asset-ref relation to asset graphql schema
b12027f8 remove singular form of asset-properties
5ef88850 adds AssetProperties tests
e37c635d Adds graphql schemas for asset-mapping
